### ISSUE TITLE
Boston Housing Example Returns Unexpected Results

### ISSUE THREAD

**Issue by PeterGrant** [EpistasisLab/tpot#1217]
Hello,

I'm a new user of TPOT running the examples to learn how to use the tool. When I run the example predicting Boston housing prices I do not receive the expected results.

I'm currently running TPOT in a Jupyter Notebook, via Anaconda with the following versions:
TPOT: 0.11.7
sklearn: 0.24.2
notebook: 6.4.0

I added the following code to my Jupyter notebook (Separately trying typing it myself, and copy/pasting the code from the example website. The version pasted here was copy/pasted from the example website):

from tpot import TPOTRegressor
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

housing = load_boston()
X_train, X_test, y_train, y_test = train_test_split(housing.data, housing.target,
                                                    train_size=0.75, test_size=0.25, random_state=42)

tpot = TPOTRegressor(generations=5, population_size=50, verbosity=2, random_state=42)
tpot.fit(X_train, y_train)
print(tpot.score(X_test, y_test))
tpot.export('tpot_boston_pipeline.py')

I receive the following output in my Jupyter Notebook:

Generation 1 - Current best internal CV score: -12.183902491076243

Generation 2 - Current best internal CV score: -12.183902491076243

Generation 3 - Current best internal CV score: -12.064332520818585

Generation 4 - Current best internal CV score: -12.064332520818585

Generation 5 - Current best internal CV score: -12.064332520818585

Best pipeline: XGBRegressor(RobustScaler(input_matrix), learning_rate=0.1, max_depth=9, min_child_weight=14, n_estimators=100, n_jobs=1, objective=reg:squarederror, subsample=0.7000000000000001, verbosity=0)
-11.070680162041242

And the following is saved in tpot_boston_pipeline.py:

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import RobustScaler
from xgboost import XGBRegressor
from tpot.export_utils import set_param_recursive

# NOTE: Make sure that the outcome column is labeled 'target' in the data file
tpot_data = pd.read_csv('PATH/TO/DATA/FILE', sep='COLUMN_SEPARATOR', dtype=np.float64)
features = tpot_data.drop('target', axis=1)
training_features, testing_features, training_target, testing_target = \
            train_test_split(features, tpot_data['target'], random_state=42)

# Average CV score on the training set was: -12.064332520818585
exported_pipeline = make_pipeline(
    RobustScaler(),
    XGBRegressor(learning_rate=0.1, max_depth=9, min_child_weight=14, n_estimators=100, n_jobs=1, objective="reg:squarederror", subsample=0.7000000000000001, verbosity=0)
)
# Fix random state for all the steps in exported pipeline
set_param_recursive(exported_pipeline.steps, 'random_state', 42)

exported_pipeline.fit(training_features, training_target)
results = exported_pipeline.predict(testing_features)

This is contrast to the expected results posted on the website, including the statement "achieves at least 10 mean squared error (MSE) on the test set", and the following code exported to tpot_boston_pipeline.py:

import numpy as np
import pandas as pd
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import PolynomialFeatures
from tpot.export_utils import set_param_recursive

# NOTE: Make sure that the outcome column is labeled 'target' in the data file
tpot_data = pd.read_csv('PATH/TO/DATA/FILE', sep='COLUMN_SEPARATOR', dtype=np.float64)
features = tpot_data.drop('target', axis=1)
training_features, testing_features, training_target, testing_target = \
            train_test_split(features, tpot_data['target'], random_state=42)

# Average CV score on the training set was: -10.812040755234403
exported_pipeline = make_pipeline(
    PolynomialFeatures(degree=2, include_bias=False, interaction_only=False),
    ExtraTreesRegressor(bootstrap=False, max_features=0.5, min_samples_leaf=2, min_samples_split=3, n_estimators=100)
)
# Fix random state for all the steps in exported pipeline
set_param_recursive(exported_pipeline.steps, 'random_state', 42)

exported_pipeline.fit(training_features, training_target)
results = exported_pipeline.predict(testing_features)

Am I doing something wrong here? Is there another package that I need to make sure is updated? Is there something wrong with the example?

Thanks,
-Peter Grant
Lawrence Berkeley National Laboratory

**Comment by rachitk** [comment]
Hi @PeterGrant! Thanks for your question and for bringing this to our attention.

The [examples](https://epistasislab.github.io/tpot/examples/) in the documentation were run using a specific version of TPOT and sklearn, and the exact pipeline exported may be different based on package versions. The purpose of declaring the random state here is for reproducibility purposes with the same environment - to make it so that you can reproduce the same pipeline and run state across different runs of TPOT if you are using the same package versions. If you would like to explore different pipelines, you can initialize TPOT with different random seeds (or omit the random_state argument entirely for an arbitrary random seed) to see what different pipelines TPOT finds at random based on how it performs mutations and generates the initial population. 

I've tested the example you posted and created a virtual environment with the same package versions, and I do indeed get the same pipeline as you. The **difference from the example isn't due to any error on your part**, and you shouldn't be concerned as the variation here appears to be due to package versions and the pipeline found by TPOT here has comparable performance on the test set to the pipeline proposed in the example (an MSE of ~11.1 compared to an MSE of ~10.8), indicating that there isn't a major regression/degradation in performance that would need investigation as packages are updated/modified.

FYI: sklearn 0.24.0 in particular made several small changes to some of the logic associated with some operators, so the final pipeline generated may be different than the versions the original example was made using (which was back in sklearn <0.22). We will look back at the examples when we get a chance and update them to better reflect what might be expected if using a version of sklearn >0.24 as well as make it clearer that the exact pipelines returned may vary depending on what versions of the dependencies are used. There are some further changes coming in sklearn 0.25 (which they will be rebranding to version 1.0.0), so we may wait until that version is released before updating the documentation fully.

Let us know if you have any other questions or any followup concerns!

**Comment by PeterGrant** [comment]
Hi @rachitk!

Great, thank you for the thorough and detailed response. Sounds to me like there's no problem here for either of us to be worried about, and this discrepancy is merely a result of sklearn being updated. Cool.

Have a good one,
-Peter
